target/linux/lantiq/patches-5.15/0746-net-dsa-lantiq_gswip-Implement-port-_pre-_bridge_fla.patch

```diff
@@ -0,0 +1,186 @@
+@@ -186,6 +188,10 @@
+ #define  GSWIP_PCE_PCTRL_0_PSTATE_LEARNING	0x3
+ #define  GSWIP_PCE_PCTRL_0_PSTATE_FORWARDING	0x7
+ #define  GSWIP_PCE_PCTRL_0_PSTATE_MASK	GENMASK(2, 0)
++#define GSWIP_PCE_PCTRL_1p(p)		(0x481 + ((p) * 0xA))
++#define  GSWIP_PCE_PCTRL_1_LRNLIM	GENMASK(7, 0)
++#define GSWIP_PCE_PCTRL_3p(p)		(0x483 + ((p) * 0xA))
++#define  GSWIP_PCE_PCTRL_3_LNDIS	BIT(15) /* Learning Disable */
+ #define GSWIP_PCE_VCTRL(p)		(0x485 + ((p) * 0xA))
+ #define  GSWIP_PCE_VCTRL_UVR		BIT(0)	/* Unknown VLAN Rule */
+ #define  GSWIP_PCE_VCTRL_VIMR		BIT(3)	/* VLAN Ingress Member violation rule */
+@@ -621,6 +627,40 @@ static int gswip_pce_table_entry_write(s
+ 				      GSWIP_PCE_TBL_CTRL_BAS);
+ }
+ 
++static void gswip_port_set_learning(struct gswip_priv *priv, int port,
++				    bool enable)
++{
++	/* Learning limit values:
++	 * - 0x0: disabled, no learning
++	 * - 0x1..0xfe: learning is limited up to 1..254 entries
++	 * - 0xff = unlimited (until the global table is full)
++	 */
++	gswip_switch_mask(priv, GSWIP_PCE_PCTRL_1_LRNLIM,
++			  enable ? GSWIP_PCE_PCTRL_1_LRNLIM : 0,
++			  GSWIP_PCE_PCTRL_1p(port));
++
++	/* GSWIP 2.2 (GRX300) and later has a dedicated learning disable bit */
++	gswip_switch_mask(priv, GSWIP_PCE_PCTRL_3_LNDIS,
++			  enable ? 0 : GSWIP_PCE_PCTRL_3_LNDIS,
++			  GSWIP_PCE_PCTRL_3p(port));
++}
++
++static void gswip_port_set_multicast_broadcast_flood(struct gswip_priv *priv,
++						     int port, bool enable)
++{
++	/* DMCPMAP: Default Multicast Port Map (also used for broadcasts) */
++	gswip_switch_mask(priv, BIT(port), enable ? BIT(port) : 0,
++			  GSWIP_PCE_PMAP2);
++}
++
++static void gswip_port_set_unicast_flood(struct gswip_priv *priv, int port,
++					 bool enable)
++{
++	/* UUCMAP: Default Unknown Unicast Port Map */
++	gswip_switch_mask(priv, BIT(port), enable ? BIT(port) : 0,
++			  GSWIP_PCE_PMAP3);
++}
++
+ /* Add the LAN port into a bridge with the CPU port by
+  * default. This prevents automatic forwarding of
+  * packages between the LAN ports when no explicit
+@@ -753,6 +793,47 @@ static int gswip_pce_load_microcode(stru
+ 	return 0;
+ }
+ 
++static int gswip_port_pre_bridge_flags(struct dsa_switch *ds, int port,
++				       struct switchdev_brport_flags flags,
++				       struct netlink_ext_ack *extack)
++{
++	bool broadcast_flood = !!(flags.val & BR_BCAST_FLOOD);
++	bool multicast_flood = !!(flags.val & BR_MCAST_FLOOD);
++
++	if (flags.mask & ~(BR_LEARNING | BR_FLOOD | BR_MCAST_FLOOD |
++			   BR_BCAST_FLOOD))
++		return -EINVAL;
++
++	if (broadcast_flood != multicast_flood) {
++		NL_SET_ERR_MSG_MOD(extack,
++				   "This chip cannot configure multicast flooding independently of broadcast flooding");
++			return -EINVAL;
++	}
++
++	return 0;
++}
++
++static int gswip_port_bridge_flags(struct dsa_switch *ds, int port,
++				   struct switchdev_brport_flags flags,
++				   struct netlink_ext_ack *extack)
++{
++	struct gswip_priv *priv = ds->priv;
++
++	if (flags.mask & BR_LEARNING)
```

Comment:
Please note that this breaks kernel selftests and I think there are some real world side-effects.
It's been a while, but with Linux 6.1 the following patches should work once you're on Linux 6.1:
- https://github.com/xdarklight/openwrt-packages/commit/ffea5021e88dcd7aba01e58043744c7d58f4035f
- https://github.com/xdarklight/openwrt-packages/commit/27af84dcf39a29c2625be3e95cff6ba19465cf27

Reply:
I can drop this patch for now.

There are a lot of changes in DSA between kernel 5.15 and 6.1. I am fine if the gswip DSA driver has the same bugs and functionality with kernel 5.15 we had with kernel 5.10.